mat64/qr.go

```diff
@@ -7,179 +7,211 @@ package mat64
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack/lapack64"
 )
 
-type QRFactor struct {
-	QR    *Dense
-	rDiag []float64
+// QR is a type for creating and using the QR factorization of a matrix.
+type QR struct {
+	qr  *Dense
+	tau []float64
 }
 
-// QR computes a QR Decomposition for an m-by-n matrix a with m >= n by Householder
-// reflections, the QR decomposition is an m-by-n orthogonal matrix q and an n-by-n
-// upper triangular matrix r so that a = q.r. QR will panic with ErrShape if m < n.
+// Factorize computes the QR factorization of an m×n matrix a where m >= n. The QR
+// factorization always exists even if A is singular.
 //
-// The QR decomposition always exists, even if the matrix does not have full rank,
-// so QR will never fail unless m < n. The primary use of the QR decomposition is
-// in the least squares solution of non-square systems of simultaneous linear equations.
-// This will fail if QRIsFullRank() returns false. The matrix a is overwritten by the
-// decomposition.
-func QR(a *Dense) QRFactor {
-	// Initialize.
+// The QR decomposition is a factorization of the matrix A such that A = Q * R.
+// The matrix Q is an orthonormal m×m matrix, and R is an m×n upper triangular matrix.
+// Q and R can be extracted from the QFromQR and RFromQR methods on Dense.
+func (qr *QR) Factorize(a Matrix) {
 	m, n := a.Dims()
 	if m < n {
 		panic(ErrShape)
 	}
-
-	qr := a
-	rDiag := make([]float64, n)
-
-	// Main loop.
-	for k := 0; k < n; k++ {
-		// Compute 2-norm of k-th column without under/overflow.
-		var norm float64
-		for i := k; i < m; i++ {
-			norm = math.Hypot(norm, qr.at(i, k))
-		}
-
-		if norm != 0 {
-			// Form k-th Householder vector.
-			if qr.at(k, k) < 0 {
-				norm = -norm
-			}
-			for i := k; i < m; i++ {
-				qr.set(i, k, qr.at(i, k)/norm)
-			}
-			qr.set(k, k, qr.at(k, k)+1)
-
-			// Apply transformation to remaining columns.
-			for j := k + 1; j < n; j++ {
-				var s float64
-				for i := k; i < m; i++ {
-					s += qr.at(i, k) * qr.at(i, j)
-				}
-				s /= -qr.at(k, k)
-				for i := k; i < m; i++ {
-					qr.set(i, j, qr.at(i, j)+s*qr.at(i, k))
-				}
-			}
-		}
-		rDiag[k] = -norm
+	k := min(m, n)
+	if qr.qr == nil {
+		qr.qr = &Dense{}
 	}
+	qr.qr.Clone(a)
+	work := make([]float64, 1)
+	qr.tau = make([]float64, k)
+	lapack64.Geqrf(qr.qr.mat, qr.tau, work, -1)
 
-	return QRFactor{qr, rDiag}
+	work = make([]float64, int(work[0]))
+	lapack64.Geqrf(qr.qr.mat, qr.tau, work, len(work))
 }
 
-// IsFullRank returns whether the R matrix and hence a has full rank.
-func (f QRFactor) IsFullRank() bool {
-	for _, v := range f.rDiag {
-		if v == 0 {
-			return false
+// TODO(btracey): Add in the "Reduced" forms for extracting the n×n orthogonal
+// and upper triangular matrices.
+
+// RFromQR extracts the m×n upper trapezoidal matrix from a QR decomposition.
+func (m *Dense) RFromQR(qr *QR) {
+	r, c := qr.qr.Dims()
+	m.reuseAs(r, c)
+	for i := 0; i < r; i++ {
+		for j := 0; j < min(i, c); j++ {
```

Comment:
So let's leave this for now and we can modify when the fast path is in.
On Aug 12, 2015 6:50 PM, "Dan Kortschak" <notifications@github.com> wrote:

> In mat64/qr.go
> <https://github.com/gonum/matrix/pull/184#discussion_r36931895>:
>
> >  }
> >
> > -// IsFullRank returns whether the R matrix and hence a has full rank.
> > -func (f QRFactor) IsFullRank() bool {
> > -	for _, v := range f.rDiag {
> > -		if v == 0 {
> > -			return false
> > +// TODO(btracey): Add in the "Reduced" forms for extracting the n×n orthogonal
> > +// and upper triangular matrices.
> > +
> > +// RFromQR extracts the m×n upper trapezoidal matrix from a QR decomposition.
> > +func (m *Dense) RFromQR(qr *QR) {
> > +	r, c := qr.qr.Dims()
> > +	m.reuseAs(r, c)
> > +	for i := 0; i < r; i++ {
> > +		for j := 0; j < min(i, c); j++ {
>
> but then again maybe I've been living in lapack too much
>
> Maybe.
>
> If you have a call m.Copy(asTriDense(qr)) it's pretty obvious what is
> happening. Though we should add the triangular cases to Dense.Copy since
> they currently get handled as the slow path.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/matrix/pull/184/files#r36931895>.
>